cmd/unifyLabels.go

```diff
@@ -0,0 +1,209 @@
+}
+
+func checkForMissingLabels(repo string, githubClient *github.Client, labelDefinitions []*LabelDesc, labelModifier LablesModifier) {
+	labels := listLabels(repo, githubClient)
+	for _, label := range labels {
+		labelDesc := findLabelDefinitionByName(*label.Name, labelDefinitions)
+		if labelDesc == nil {
+			labelDescByOldName := findLabelDefinitionByOldName(*label.Name, labelDefinitions)
+			if labelDescByOldName == nil {
+				labelModifier.removeLabel(label)
+			} else {
+				labelModifier.renameLabel(label, labelDescByOldName)
+			}
+		}
+	}
+}
+
+func checkExistingLabels(repo string, githubClient *github.Client, labelDefinitions []*LabelDesc, labelModifier LablesModifier) {
+	labels := listLabels(repo, githubClient) // list again to get renamed
+	for _, labelDefinition := range labelDefinitions {
+		label := findLabelByName(labelDefinition.name, labels)
+		if label == nil {
+			if labelDefinition.required {
+				labelModifier.createLabel(labelDefinition)
+			}
+		} else {
+			if *label.Color != labelDefinition.color {
+				labelModifier.setColor(label, labelDefinition)
+			}
+		}
+	}
+}
+
+func listLabels(repo string, githubClient *github.Client) []*github.Label {
+	labels, _, err := githubClient.Issues.ListLabels(context.Background(), "exasol", repo, &github.ListOptions{PerPage: 100})
+	if err != nil {
+		panic("Failed to list labels")
+	}
+	return labels
+}
+
+type LablesModifier interface {
+	createLabel(labelDefinition *LabelDesc)
+	removeLabel(label *github.Label)
+	renameLabel(label *github.Label, labelDefinition *LabelDesc)
+	setColor(label *github.Label, labelDefinition *LabelDesc)
+}
+
+type DryRunLabelModifier struct {
+}
+
+func (dryRunModifier *DryRunLabelModifier) createLabel(labelDefinition *LabelDesc) {
+	fmt.Printf("Missing required label '%s'. Would create.\n", labelDefinition.name)
+}
+
+func (dryRunModifier *DryRunLabelModifier) removeLabel(label *github.Label) {
+	fmt.Printf("Superfluous label '%s'. Would remove.\n", *label.Name)
+}
+
+func (dryRunModifier *DryRunLabelModifier) renameLabel(label *github.Label, labelDefinition *LabelDesc) {
+	fmt.Printf("The label '%s' was renamed to '%s'. Would rename.\n", *label.Name, labelDefinition.name)
+}
+
+func (dryRunModifier *DryRunLabelModifier) setColor(label *github.Label, labelDefinition *LabelDesc) {
+	fmt.Printf("Label '%s' has wrong color %s. Expected: %s. Would change.\n", *label.Name, *label.Color, labelDefinition.color)
+}
+
+type RealLabelModifier struct {
+	githubClient *github.Client
+	repo         string
+}
+
+func (realRunModifer *RealLabelModifier) createLabel(labelDefinition *LabelDesc) {
+	_, _, err := realRunModifer.githubClient.Issues.CreateLabel(context.Background(), "exasol", realRunModifer.repo, &github.Label{Name: &labelDefinition.name, Color: &labelDefinition.color})
+	if err != nil {
+		panic(fmt.Sprintf("Failed to create label '%s' for repo '%s'. Cause: '%s'", labelDefinition.name, realRunModifer.repo, err.Error()))
```

Comment:
is panic really the correct reaction, that means we can't continue with anything else, or is the assumption if it fails, rerun the program